app/com/arpnetworking/metrics/util/PagingIterator.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2019 Dropbox, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.metrics.util;
+
+import com.arpnetworking.commons.builder.OvalBuilder;
+import net.sf.oval.constraint.NotNull;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+/**
+ * Iterator that wraps a paging function, turning a "yield batches of elements at once" interface into a plain {@link Iterator}.
```

Comment:
This works for back-end where you can store the iterator instance. The other change we talked about which link the Query and QueryResult to allow for one to continue from the other would be helpful to API/frontend interactions. I wonder if that would simplify the iterator as well?

Reply:
That would simplify this! I was imagining refactoring a common `Query` interface out of `HostQuery`, `AlertQuery`, and `ReportQuery`, and then writing a method somewhere of a type like `Query<E> -> Iterator<QueryResult<E>>`. But this ends up requiring a nontrivial refactor to deal with slight differences between the `...Query` classes, and... I didn't think it was worth it at this time, given that (in my opinion?) this `PagingIterator` class is conceptually simple and is only slightly less elegant to use.

I could definitely go down the other path, though. I didn't see anything that looked like a dealbreaker, it just seemed harder.